doc/README.twom.md

```diff
@@ -0,0 +1,189 @@
+
+## The twoskip file format
+
+Twoskip added an extra innovation to the skiplist format - a second linked list at the bottom level, which includes every single record in order - but with a twist.  The two pointers leapfrog each other, so it’s always the lowest pointer which is updated each time.  This means that if there’s a crash, the pointer which doesn’t point past the old size of the file will always point to the record which used to be the next in the valid records - so a repair is just a matter of walking all the valid level 0 pointers, and zeroing out any future pointers plus fixing up the higher level  pointers.  This is so fast that transaction `abort()` is implemented as just running recovery on the file.
+
+Along with a three phase commit, this makes the twoskip file format incredibly robust.  The three phases are:
+
+1. before writing any new record to the end of file, mark the header as DIRTY, and fsync
+2. once all new records are written, fsync the entire file (still marked dirtry)
+3. mark the header NOT-DIRTY along with the new file length, and fsync
+
+This means that so long as the final header write is atomic, transactions are always completely safe.
+
+## Twoskip downsides:
+
+- Three fsyncs per transaction is kinda expensive!
+- Every write causes the file to be extended, as twoskip format uses exact file size and doesn’t allocate slop.
+- A single write does amplify into 1+N (the level of the record) writes; the append of the new record, and the updates of the immediately previous records in the skiplist at each level.  Since half the records are level 1, half that level 2, etc - the average write updates three separate previous records.  In a recently repacked file, the previous records are likely to be adjacent or nearly so for the lower levels, but this becomes less so over time, so repacks are valuable for both data locality and size reduction.
+- The repack operation requires the database to be remain exclusively locked for the duration, which is slow on large files (particularly along with the above lack of slop).
+
+## The Twom format
+
+Why the name?  It’s twoskip, but with mmap and MVCC.  Two Ms.  And the “tombstone” records, rather than just causing the skiplist to be rewritten without the deleted record in it, remain in the linked lists right until a repack, with an ANCESTOR pointer back to the previous entity with that key.
+
+This tombstone allows the MVCC transaction, because you can walk the current lowest level linked list, and any time you find a record with an offset in the future of the file size at the time you started the transaction, walk back up the ancestor chain (which is a single linked list backward through the file, to the original ADD record which was the first time the key was seen in this file).
```

Comment:
I would only walk the ancestor chain until the file offset of the record is less than then file size of the time I started the transaction.

Reply:
yes, that's what the code does - but you can create a transaction at any previous commit and read the state that it was at that commit.